### PR TITLE
keripy 1.1.x -> 1.2.x migration SignifyGroupHab fix

### LINKED ISSUE #257: Internal Server Error on connect after upgrade from 0.1.2 to 0.2.0-dev2 when agent contains group hab
I am having issues upgrading a keria instance from 0.1.2 to 0.2.0-dev2. When the instance contains an agent with a group hab, the controller cannot connect to it.

See reproduction here: https://github.com/lenkan/keria-upgrade-test. It should be possible to run it by cloning the repo and 

```
git clone git@github.com:lenkan/keria-upgrade-test.git
cd keria-upgrade-test
./run-test.sh
```

The error output is:

```
connect-1  | /app/scripts-connect/node_modules/signify-ts/src/keri/app/controller.ts:152
connect-1  |         if (state == null || state['ee']['s'] == 0) {
connect-1  |                                        ^
connect-1  |
connect-1  |
connect-1  | TypeError: Cannot read properties of undefined (reading 's')
connect-1  |     at Controller (/app/scripts-connect/node_modules/signify-ts/src/keri/app/controller.ts:152:40)
connect-1  |     at SignifyClient.connect (/app/scripts-connect/node_modules/signify-ts/src/keri/app/clienting.ts:133:27)
connect-1  |     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
connect-1  |     at <anonymous> (/app/scripts-connect/src/main.ts:13:1)
connect-1  |
connect-1  | Node.js v20.12.2
connect-1  | npm ERR! Lifecycle script `start` failed with error:
connect-1  | npm ERR! Error: command failed
connect-1  | npm ERR!   in workspace: undefined
connect-1  | npm ERR!   at location: /app/scripts-connect
connect-1 exited with code 1
keria-1  | /keripy/src/keri/core/serdering.py:120: SyntaxWarning: invalid escape sequence '\A'
keria-1  |   """Design Notes:
keria-1  | The Agency is loaded and waiting for requests...
keria-1  | 2024-06-26 13:21:20 [FALCON] [ERROR] GET /agent/EEFCS78eJySGGx9Nl7r-eHRSDZ4xbisWwkqrwEWft-bG => Traceback (most recent call last):
keria-1  |   File "falcon/app.py", line 365, in falcon.app.App.__call__
keria-1  |   File "/usr/local/var/keria/src/keria/app/aiding.py", line 110, in on_get
keria-1  |     agent = self.agency.get(caid)
keria-1  |             ^^^^^^^^^^^^^^^^^^^^^
keria-1  |   File "/usr/local/var/keria/src/keria/app/agenting.py", line 246, in get
keria-1  |     agentHby = habbing.Habery(name=caid, base=self.base, bran=self.bran, ks=ks, temp=self.temp)
keria-1  |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
keria-1  |   File "/keripy/src/keri/app/habbing.py", line 241, in __init__
keria-1  |     self.setup(**self._inits)  # finish setup later
keria-1  |     ^^^^^^^^^^^^^^^^^^^^^^^^^
keria-1  |   File "/keripy/src/keri/app/habbing.py", line 316, in setup
keria-1  |     self.loadHabs()
keria-1  |   File "/keripy/src/keri/app/habbing.py", line 345, in loadHabs
keria-1  |     hab = SignifyGroupHab(ks=self.ks, db=self.db, cf=self.cf, mgr=self.mgr,
keria-1  |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
keria-1  | TypeError: SignifyGroupHab.__init__() missing 1 required positional argument: 'smids'
keria-1  |
```

Could it be that the migration is not done correctly? If so, does anyone know how to correctly upgrade all the databases in a keria instance? Right now it does `kli migrate run --name <agent aid>` for all directories inside /usr/local/var/keri/db. See https://github.com/lenkan/keria-upgrade-test/blob/main/migrate.sh

Note that the upgrade and connection works if the agent does not contain a group.

### PR DESCRIPTION
**Pull request by m00sey**
In conjunction with @lenkan 's original test repo, cc @iFergal .

https://github.com/m00sey/keria-upgrade-test/tree/sig-fix

```
Hby EEFCS78eJySGGx9Nl7r-eHRSDZ4xbisWwkqrwEWft-bG

	 group1 - EC6hdfWAszbItZN9AMHF3hRqt5PDAaGsIK8858D9jT2a - <class 'keri.app.habbing.SignifyGroupHab'>


	 Proposed smids and rmids updates for group1 - EC6hdfWAszbItZN9AMHF3hRqt5PDAaGsIK8858D9jT2a:
		 smids: {'EIJjVcJInX-ELwoQ9ZGQRKpuCEzmuqYa6Aq5zU5RqA8l', 'EGuzYjhz3tKQFhrh3HsdP2ys2v5WrXUTA3agsq-3RtS4'}
			 -> EIJjVcJInX-ELwoQ9ZGQRKpuCEzmuqYa6Aq5zU5RqA8l qar1
			 -> EGuzYjhz3tKQFhrh3HsdP2ys2v5WrXUTA3agsq-3RtS4 qar3
		 rmids: {'EIJjVcJInX-ELwoQ9ZGQRKpuCEzmuqYa6Aq5zU5RqA8l', 'EGuzYjhz3tKQFhrh3HsdP2ys2v5WrXUTA3agsq-3RtS4'}
			 -> EIJjVcJInX-ELwoQ9ZGQRKpuCEzmuqYa6Aq5zU5RqA8l qar1
			 -> EGuzYjhz3tKQFhrh3HsdP2ys2v5WrXUTA3agsq-3RtS4 qar3
	 Updating HabitatRecord(hid='EC6hdfWAszbItZN9AMHF3hRqt5PDAaGsIK8858D9jT2a', name='group1', domain=None, mid='EIJjVcJInX-ELwoQ9ZGQRKpuCEzmuqYa6Aq5zU5RqA8l', smids=None, rmids=None, sid='EC6hdfWAszbItZN9AMHF3hRqt5PDAaGsIK8858D9jT2a', watchers=[])

	 skipping ED4AQf_aVbauZvO79AGHucq5gArqbYmmbQ-lEu9mMjNz - <class 'keri.app.habbing.Hab'>
	 skipping EEFCS78eJySGGx9Nl7r-eHRSDZ4xbisWwkqrwEWft-bG - <class 'keri.app.habbing.SignifyHab'>
	 skipping EIJjVcJInX-ELwoQ9ZGQRKpuCEzmuqYa6Aq5zU5RqA8l - <class 'keri.app.habbing.SignifyHab'>

	 group0 - EK2gkA-gSASwgCnr7hHAha_3krLxLSqmsbS-YH40qRdi - <class 'keri.app.habbing.SignifyGroupHab'>


	 Proposed smids and rmids updates for group0 - EK2gkA-gSASwgCnr7hHAha_3krLxLSqmsbS-YH40qRdi:
		 smids: {'EIJjVcJInX-ELwoQ9ZGQRKpuCEzmuqYa6Aq5zU5RqA8l', 'ELhW9hbFlL6J46i6ScWZuSYBzEVpa36O62c4BCSVC0o7'}
			 -> EIJjVcJInX-ELwoQ9ZGQRKpuCEzmuqYa6Aq5zU5RqA8l qar1
			 -> ELhW9hbFlL6J46i6ScWZuSYBzEVpa36O62c4BCSVC0o7 qar2
		 rmids: {'EIJjVcJInX-ELwoQ9ZGQRKpuCEzmuqYa6Aq5zU5RqA8l', 'ELhW9hbFlL6J46i6ScWZuSYBzEVpa36O62c4BCSVC0o7'}
			 -> EIJjVcJInX-ELwoQ9ZGQRKpuCEzmuqYa6Aq5zU5RqA8l qar1
			 -> ELhW9hbFlL6J46i6ScWZuSYBzEVpa36O62c4BCSVC0o7 qar2
	 Updating HabitatRecord(hid='EK2gkA-gSASwgCnr7hHAha_3krLxLSqmsbS-YH40qRdi', name='group0', domain=None, mid='EIJjVcJInX-ELwoQ9ZGQRKpuCEzmuqYa6Aq5zU5RqA8l', smids=None, rmids=None, sid='EK2gkA-gSASwgCnr7hHAha_3krLxLSqmsbS-YH40qRdi', watchers=[])


Hby EKg4HKlaKDbTAhpFFTYVmiLCFSqjCHg6xBQa8xUMx4Se
	 skipping EJcnFAOIO087C-lq2CMxyQsWm8qy8v6_OplUhst2pb0U - <class 'keri.app.habbing.Hab'>

	 group0 - EK2gkA-gSASwgCnr7hHAha_3krLxLSqmsbS-YH40qRdi - <class 'keri.app.habbing.SignifyGroupHab'>


	 Proposed smids and rmids updates for group0 - EK2gkA-gSASwgCnr7hHAha_3krLxLSqmsbS-YH40qRdi:
		 smids: {'EIJjVcJInX-ELwoQ9ZGQRKpuCEzmuqYa6Aq5zU5RqA8l', 'ELhW9hbFlL6J46i6ScWZuSYBzEVpa36O62c4BCSVC0o7'}
			 -> EIJjVcJInX-ELwoQ9ZGQRKpuCEzmuqYa6Aq5zU5RqA8l qar1
			 -> ELhW9hbFlL6J46i6ScWZuSYBzEVpa36O62c4BCSVC0o7 qar2
		 rmids: {'EIJjVcJInX-ELwoQ9ZGQRKpuCEzmuqYa6Aq5zU5RqA8l', 'ELhW9hbFlL6J46i6ScWZuSYBzEVpa36O62c4BCSVC0o7'}
			 -> EIJjVcJInX-ELwoQ9ZGQRKpuCEzmuqYa6Aq5zU5RqA8l qar1
			 -> ELhW9hbFlL6J46i6ScWZuSYBzEVpa36O62c4BCSVC0o7 qar2
	 Updating HabitatRecord(hid='EK2gkA-gSASwgCnr7hHAha_3krLxLSqmsbS-YH40qRdi', name='group0', domain=None, mid='ELhW9hbFlL6J46i6ScWZuSYBzEVpa36O62c4BCSVC0o7', smids=None, rmids=None, sid='EK2gkA-gSASwgCnr7hHAha_3krLxLSqmsbS-YH40qRdi', watchers=[])

	 skipping EKg4HKlaKDbTAhpFFTYVmiLCFSqjCHg6xBQa8xUMx4Se - <class 'keri.app.habbing.SignifyHab'>
	 skipping ELhW9hbFlL6J46i6ScWZuSYBzEVpa36O62c4BCSVC0o7 - <class 'keri.app.habbing.SignifyHab'>

Hby ELkJ8zmKQePs6Vi7If3SGSUKH84LGs0I8Ef-QV94A_9O

	 group1 - EC6hdfWAszbItZN9AMHF3hRqt5PDAaGsIK8858D9jT2a - <class 'keri.app.habbing.SignifyGroupHab'>


	 Proposed smids and rmids updates for group1 - EC6hdfWAszbItZN9AMHF3hRqt5PDAaGsIK8858D9jT2a:
		 smids: {'EIJjVcJInX-ELwoQ9ZGQRKpuCEzmuqYa6Aq5zU5RqA8l', 'EGuzYjhz3tKQFhrh3HsdP2ys2v5WrXUTA3agsq-3RtS4'}
			 -> EIJjVcJInX-ELwoQ9ZGQRKpuCEzmuqYa6Aq5zU5RqA8l qar1
			 -> EGuzYjhz3tKQFhrh3HsdP2ys2v5WrXUTA3agsq-3RtS4 qar3
		 rmids: {'EIJjVcJInX-ELwoQ9ZGQRKpuCEzmuqYa6Aq5zU5RqA8l', 'EGuzYjhz3tKQFhrh3HsdP2ys2v5WrXUTA3agsq-3RtS4'}
			 -> EIJjVcJInX-ELwoQ9ZGQRKpuCEzmuqYa6Aq5zU5RqA8l qar1
			 -> EGuzYjhz3tKQFhrh3HsdP2ys2v5WrXUTA3agsq-3RtS4 qar3
	 Updating HabitatRecord(hid='EC6hdfWAszbItZN9AMHF3hRqt5PDAaGsIK8858D9jT2a', name='group1', domain=None, mid='EGuzYjhz3tKQFhrh3HsdP2ys2v5WrXUTA3agsq-3RtS4', smids=None, rmids=None, sid='EC6hdfWAszbItZN9AMHF3hRqt5PDAaGsIK8858D9jT2a', watchers=[])

	 skipping EGuzYjhz3tKQFhrh3HsdP2ys2v5WrXUTA3agsq-3RtS4 - <class 'keri.app.habbing.SignifyHab'>
	 skipping ELkJ8zmKQePs6Vi7If3SGSUKH84LGs0I8Ef-QV94A_9O - <class 'keri.app.habbing.SignifyHab'>
	 skipping EPCzcq3dBwrf7WPqOSBNfZq5qqPYsrv2Hp8cKuyHWHhz - <class 'keri.app.habbing.Hab'>
```

```
keria-1  | The Agency is loaded and waiting for requests...
keria-1  |   Agent: ED4AQf_aVbauZvO79AGHucq5gArqbYmmbQ-lEu9mMjNz   Controller: EEFCS78eJySGGx9Nl7r-eHRSDZ4xbisWwkqrwEWft-bG
```

Closes #257

### CODE REVIEW

**Comment by iFergal** [comment]
Nice, thank you! Works on my machine too.

I would say this works for the given issue and long term we should have a proper migration strategy on KERIA too for when `SignifyGroupHab` etc have moved (based on https://github.com/WebOfTrust/keripy/issues/824). Otherwise it's not clear to those upgrading which cli migrate/fix scripts need to be called when - ideally a `keria migrate` would migrate keripy and KERIA in a deterministic way.

**Comment by 2byrds** [comment]
> Nice, thank you! Works on my machine too.
> 
> I would say this works for the given issue and long term we should have a proper migration strategy on KERIA too for when `SignifyGroupHab` etc have moved (based on [WebOfTrust/keripy#824](https://github.com/WebOfTrust/keripy/issues/824)). Otherwise it's not clear to those upgrading which cli migrate/fix scripts need to be called when - ideally a `keria migrate` would migrate keripy and KERIA in a deterministic way.

Good stuff @m00sey! Per Fergal comment should this PR be renamed to be similar to the branch name that is being merged, to avoid confusion? Something like "keripy 1.1.x -> 1.2.x migration SignifyGroupHab fix"?
- Based on this SignifyGroupHab migration fix
- And based on @iFergal comments and related to his [migrate work](https://github.com/WebOfTrust/keria/pull/272)
- and [the recent discussion about migration on discord](https://discord.com/channels/1148629222647148624/1148629228896649218/1275881125226614808),  It's progress towards
- The [community discussion this coming Tuesday](https://github.com/WebOfTrust/keri/discussions/96#discussion-7080484)
- And the medium-term plan to [move Signify related code from keripy](https://github.com/WebOfTrust/keripy/issues/824).

We are collecting a good set of tools/fixes/guidance to help community members decide when they are ready to migrate.

File: src/keria/app/cli/commands/sig-fix.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+from keria.db import basing as abase
+
+logger = help.ogler.getLogger()
+
+parser = argparse.ArgumentParser(description='Copy HabitatRecord smid and rmid data')
+parser.set_defaults(handler=lambda args: handler(args),
+                    transferable=True)
+parser.add_argument('--base', '-b', help='additional optional prefix to file location of KERI keystore',
+                    required=False, default="")
+parser.add_argument('--force', action="store_true", required=False, default=False,
```
Comment (2byrds) [inline src/keria/app/cli/commands/sig-fix.py:25]: I like that the user has to explicitly force it to make changes

**Comment by lenkan** [comment]
🚀 

@m00sey or @2byrds are you OK with pushing a new 0.2.0-devX docker tag with this fix? Thank you!

**Comment by lenkan** [comment]
FYI - I have update the upgrade test repo with these changes and simplified it a little bit: https://github.com/lenkan/keria-upgrade-test

When do you think you can push the new docker tag?

**Comment by 2byrds** [comment]
@lenkan I don't think I can push an image to WebOfTrust docker. Perhaps @m00sey can? I would support a keria github action that will push a new dev image on every PR merged to main. Similar to:

- https://github.com/GLEIF-IT/vlei-verifier/blob/main/.github/workflows/publish-verifier.yml

- https://hub.docker.com/r/gleif/vlei-verifier/tags

**Pull request merged**